File: android/app/src/main/java/com/bwqr/mavinote/reax/Runtime.kt
package com.bwqr.mavinote.reax

import android.util.Log
import com.bwqr.mavinote.BuildConfig
import com.bwqr.mavinote.models.NoteError
import com.novi.bincode.BincodeDeserializer
import com.novi.serde.DeserializationError
import kotlinx.coroutines.CancellableContinuation
import kotlinx.coroutines.cancel
import kotlinx.coroutines.channels.ProducerScope
import kotlinx.coroutines.channels.awaitClose
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.callbackFlow
import kotlinx.coroutines.suspendCancellableCoroutine
import java.util.concurrent.Semaphore
import kotlin.concurrent.thread
import kotlin.coroutines.resume
import kotlin.coroutines.resumeWithException
import kotlin.random.Random

typealias AsyncStart = (Int) -> Long
typealias SyncStart = () -> ByteArray

interface Future {
    // Return value indicates whether the future is completed or not
    fun handle(bytes: ByteArray): Boolean

    fun abort()
}

class Stream<T> constructor(
    private val scope: ProducerScope<T>,
    private val deserialize: Deserialize<T>,
) : Future {
    private var joinHandle: Long? = null

    override fun handle(bytes: ByteArray): Boolean {
        val deserializer = BincodeDeserializer(bytes)

        return when (deserializer.deserialize_variant_index()) {
            0 -> {
                when (val either = Either.Deserialize(deserialize, NoteError).deserialize(deserializer)) {
                    is Either.Success -> scope.trySend(either.value)
                    is Either.Failure -> scope.cancel("", either.value)
                }

                false
            }

            1 -> {
                scope.close()
                true
            }

            else -> throw DeserializationError("Unknown variant index for Stream")
        }
    }

    override fun abort() {
        val joinHandle = joinHandle ?: throw Error("A Stream without a joinHandle is being aborted")

        _abort(joinHandle)
    }

    fun setJoinHandle(handle: Long) {
        joinHandle = handle
    }
}

class Once<T> constructor(
    private val continuation: CancellableContinuation<T>,
    private val deserialize: Deserialize<T>,
) : Future {
    private var joinHandle: Long? = null

    override fun handle(bytes: ByteArray): Boolean {
        val deserializer = BincodeDeserializer(bytes)

        when (val either = Either.Deserialize(deserialize, NoteError).deserialize(deserializer)) {
            is Either.Success -> continuation.resume(either.value)
            is Either.Failure -> continuation.resumeWithException(either.value)
        }

        return true
    }

    override fun abort() {
        val joinHandle = joinHandle ?: throw Error("A Once without a joinHandle is being aborted")

        _abort(joinHandle)
    }

    fun setJoinHandle(handle: Long) {
        joinHandle = handle
    }
}

private class CriticalSection<T> constructor(private val instance: T) {
    private val semaphore: Semaphore = Semaphore(1)

    fun <R> enter(callback: (T) -> R): R {
        semaphore.acquire()
        val res = callback(instance)
        semaphore.release()

        return res
    }
}

class Runtime private constructor() {
    private var futures: CriticalSection<HashMap<Int, Future>> = CriticalSection(HashMap())

    companion object {
        private lateinit var instance: Runtime

        // Return value indicates if the Runtime is already initialized
        fun init(filesDir: String): Either<Boolean, String> {
            if (this::instance.isInitialized) {
                return Either.Success(true)
            }

            System.loadLibrary("reax")

            val either = run(Either.Deserialize(DeUnit, DeString)) {
                _init(
                    BuildConfig.API_URL,
                    BuildConfig.WS_URL,
                    filesDir
                )
            }

            if (either is Either.Failure) {
                return Either.Failure(either.value)
            }

            instance = Runtime()

            thread {
                _initHandler { id, bytes ->
                    Log.d(
                        "Runtime",
                        "received message with id:$id byteLength:${bytes.size}"
                    )

                    val future = instance.futures.enter { futures ->
                        futures[id] ?: throw Error("A message with unknown id is received $id")
                    }

                    if (future.handle(bytes)) {
                        instance.abort(id)
                    }
                }
            }

            return Either.Success(false)
        }

        fun <T> runStream(deserialize: Deserialize<T>, onStart: AsyncStart): Flow<T> =
            instance.runStream(deserialize, onStart)

        suspend fun runOnceUnit(onStart: AsyncStart) = instance.runOnce(DeUnit, onStart)

        suspend fun <T> runOnce(deserialize: Deserialize<T>, onStart: AsyncStart): T =
            instance.runOnce(deserialize, onStart)

        fun <T> run(deserialize: Deserialize<T>, onStart: SyncStart): T {
            return deserialize.deserialize(BincodeDeserializer(onStart()))
        }
    }

    private fun <T> runStream(deserialize: Deserialize<T>, onStart: AsyncStart): Flow<T> =
        callbackFlow {
            val stream = Stream(this, deserialize)

            val id = insertFuture(stream)

            stream.setJoinHandle(onStart(id))

            awaitClose {
                instance.abort(id)
            }
        }

    private suspend fun <T> runOnce(deserialize: Deserialize<T>, onStart: AsyncStart): T =
        suspendCancellableCoroutine { cont ->
            val once = Once(cont, deserialize)

            val id = insertFuture(once)

            once.setJoinHandle(onStart(id))

            cont.invokeOnCancellation {
                abort(id)
            }
        }

    private fun insertFuture(future: Future): Int {
        return futures.enter { futures ->
            var id = Random.nextInt()

            while (futures.contains(id)) {
                id = Random.nextInt()
            }

            futures[id] = future

            id
        }
    }

    private fun abort(id: Int) {
        val future = futures.enter { futures ->
            futures.remove(id) ?: throw Error("Aborting an unknown future $id")
        }

        future.abort()
    }
}

private external fun _init(apiUrl: String, wsUrl: String, storageDir: String): ByteArray
private external fun _initHandler(callback: (streamId: Int, bytes: ByteArray) -> Unit)
private external fun _abort(joinHandle: Long)

File: ios/Mavinote/Reax/Runtime.swift
import Foundation
import Serde

typealias AsyncStart = (_ id: Int32) -> UnsafeMutableRawPointer
typealias SyncStart = (_ deserializer: @convention(c) (UnsafePointer<UInt8>?, UInt) -> UnsafeMutableRawPointer?) -> UnsafeMutableRawPointer?

protocol Future {
    // Return value indicates whether the future is completed or not
    func handle(_ bytes: [UInt8]) -> Bool

    func abort()
}

class Stream<T: Deserialize>: Future {
    private let continuation: AsyncStream<T>.Continuation
    private var joinHandle: UnsafeMutableRawPointer?

    init(continuation: AsyncStream<T>.Continuation) {
        self.continuation = continuation
    }

    func handle(_ bytes: [UInt8]) -> Bool {
        let deserializer = BincodeDeserializer(input: bytes)

        do {
            switch try deserializer.deserialize_variant_index() {
            case 0:
                continuation.yield(try T.deserialize(deserializer))
                return false
            case 1:
                continuation.finish()
                return true
            default: throw DeserializationError.invalidInput(issue: "Unknown variant index for Stream Message")
            }
        } catch {
            fatalError("failed to deserialize in Stream \(error)")
        }
    }

    func abort() {
        guard let joinHandle = joinHandle else {
            fatalError("A Stream without a joinHandle is being aborted")
        }

        reax_abort(joinHandle)
    }

    func setJoinHandle(handle: UnsafeMutableRawPointer) {
        joinHandle = handle
    }
}

class Once<T: Deserialize>: Future {
    private var continuation: CheckedContinuation<T, Never>?
    private var joinHandle: UnsafeMutableRawPointer?

    func handle(_ bytes: [UInt8]) -> Bool {
        guard let continuation = continuation else {
            fatalError("A Once without a continuation is being handled")
        }

        let deserializer = BincodeDeserializer(input: bytes)

        do {
            continuation.resume(returning: try T.deserialize(deserializer))
        } catch {
            fatalError("Failed to deserialize in Once \(error)")
        }

        return true
    }


    func abort() {
        guard let joinHandle = joinHandle else {
            fatalError("A Once without a joinHandle is being aborted")
        }

        reax_abort(joinHandle)
    }

    func setJoinHandle(handle: UnsafeMutableRawPointer) {
        joinHandle = handle
    }

    func setContinuation(cont: CheckedContinuation<T, Never>) {
        continuation = cont
    }
}

private class CriticalSection<T> {
    private var instance: T
    private let semaphore = DispatchSemaphore(value: 1)

    init(instance: T) {
        self.instance = instance
    }

    func enter<R>(_ callback: (inout T) -> R) -> R {
        semaphore.wait()
        let res = callback(&instance)
        semaphore.signal()
        return res
    }
}

class Runtime {
    private static var _instance: Runtime?

    private var futures: CriticalSection<[Int32: any Future]> = CriticalSection(instance: [:])

    static func initialize(storageDir: String) -> Result<(), String> {
        if (_instance != nil) {
            return .success(())
        }

        let result: Result<DeUnit, String> = run { reax_init(API_URL, WS_URL, storageDir, $0) }

        if case .failure(let failure) = result {
            return .failure(failure)
        }

        _instance = Runtime()

        Thread
            .init(target: _instance!, selector: #selector(initHandler), object: nil)
            .start()

        return .success(())
    }

    static func run<T: Deserialize, E: Deserialize>(_ onStart: SyncStart) -> Result<T, E> {
        let ptr = onStart { bytes, byteLen in
            let array = Array(UnsafeBufferPointer(start: bytes, count: Int(byteLen)))

            return UnsafeMutableRawPointer(Unmanaged.passRetained(BincodeDeserializer(input: array)).toOpaque())
        }

        let deserializer = Unmanaged<BincodeDeserializer>.fromOpaque(ptr!).takeRetainedValue()

        do {
            let index = try deserializer.deserialize_variant_index()

            switch index {
            case 0: return .success(try T.deserialize(deserializer))
            case 1: return .failure(try E.deserialize(deserializer))
            default: throw DeserializationError.invalidInput(issue: "Unknown variant for Result \(index)")
            }
        } catch {
            fatalError("Failed to deserialize sync function \(error)")
        }
    }

    static func runStream<T: Deserialize>(_ onStart: AsyncStart) -> AsyncStream<Result<T, NoteError>> {
        return Self.instance().runStream(onStart)
    }

    static func runOnceUnit(_ onStart: AsyncStart) async -> Result<(), NoteError> {
        let res: Result<DeUnit, NoteError> = await Self.runOnce(onStart)

        return res.map { _ in }
    }

    static func runOnce<T: Deserialize>(_ onStart: AsyncStart) async -> Result<T, NoteError> {
        return await Self.instance().runOnce(onStart)
    }

    private static func instance() -> Runtime {
        _instance!
    }

    private func runStream<T: Deserialize>(_ onStart: AsyncStart) -> AsyncStream<T> {
        return AsyncStream { continuation in
            let stream = Stream<T>(continuation: continuation)
            let id = insertFuture(future: stream)

            stream.setJoinHandle(handle: onStart(id))

            continuation.onTermination = { @Sendable _ in
                self.abort(id)
            }
        }
    }

    private func runOnce<T: Deserialize>(_ onStart: AsyncStart) async -> T {
        let once = Once<T>()
        let id = insertFuture(future: once)

        return await withTaskCancellationHandler(
            operation: {
                return await withCheckedContinuation { cont in
                    once.setContinuation(cont: cont)
                    once.setJoinHandle(handle: onStart(id))
                }
            },
            onCancel: { abort(id) }
        )
    }

    private func insertFuture(future: any Future) -> Int32 {
        return futures.enter { futures in
            var id = Int32.random(in: 0...Int32.max)

            while futures[id] != nil {
                id = Int32.random(in: 0...Int32.max)
            }

            futures[id] = future

            return id
        }
    }

    private func abort(_ id: Int32) {
        guard let future = futures.enter({ futures in futures.removeValue(forKey: id) }) else {
            return
        }

        future.abort()
    }

    @objc private func initHandler() {
        let this = UnsafeMutableRawPointer(Unmanaged.passRetained(self).toOpaque())

        reax_init_handler(this) { id, bytes, bytesLen, ptr in
            let this = Unmanaged<AnyObject>.fromOpaque(ptr!).takeUnretainedValue() as! Runtime

            var bytesArray: [UInt8] = Array(repeating: 0, count: Int(bytesLen))

            for i in 0..<Int(bytesLen) {
                bytesArray[i] = bytes!.advanced(by: i).pointee
            }

            let future = this.futures.enter { futures in
                guard let future = futures[id] else {
                    fatalError("A message with unknown id is received, \(id)")
                }

                return future
            }

            if future.handle(bytesArray) {
                this.abort(id)
            }
        }

        let _ = Unmanaged<AnyObject>.fromOpaque(this).takeRetainedValue()
    }
}
